crates/miden-block-prover/src/local_block_prover.rs

```diff
@@ -0,0 +1,265 @@
+    /// - the nullifier witnesses provided in the proposed block result in a different nullifier
+    ///   tree root than the contained previous block header commits to.
+    pub fn prove(&self, proposed_block: ProposedBlock) -> Result<ProvenBlock, ProvenBlockError> {
+        self.prove_without_verification_inner(proposed_block)
+    }
+
+    /// Proves the provided [`ProposedBlock`] into a [`ProvenBlock`], **without verifying batches
+    /// and proving the block**.
+    ///
+    /// This is exposed for testing purposes.
+    #[cfg(any(feature = "testing", test))]
+    pub fn prove_without_verification(
+        &self,
+        proposed_block: ProposedBlock,
+    ) -> Result<ProvenBlock, ProvenBlockError> {
+        self.prove_without_verification_inner(proposed_block)
+    }
+
+    /// Proves the provided [`ProposedBlock`] into a [`ProvenBlock`].
+    ///
+    /// The assumptions of this method are that the checks made by construction of a
+    /// [`ProposedBlock`] are enforced.
+    ///
+    /// See [`Self::prove`] for more details.
+    fn prove_without_verification_inner(
+        &self,
+        proposed_block: ProposedBlock,
+    ) -> Result<ProvenBlock, ProvenBlockError> {
+        // Get the block number and timestamp of the new block and compute the tx commitment.
+        // --------------------------------------------------------------------------------------------
+
+        let block_num = proposed_block.block_num();
+        let timestamp = proposed_block.timestamp();
+        let tx_hash = BlockHeader::compute_tx_commitment(proposed_block.affected_accounts());
+
+        // Split the proposed block into its parts.
+        // --------------------------------------------------------------------------------------------
+
+        // TODO: We should include the batches in the block, right? If so, should we transform them
+        // into a `VerifiedBatch` struct, with the (eventually present) ZK proof removed and
+        // the batch note trees updated after erased notes on the block level have been
+        // taken into account?
+        let (
+            _batches,
+            mut account_updated_witnesses,
+            output_note_batches,
+            block_note_tree,
+            created_nullifiers,
+            chain_mmr,
+            prev_block_header,
+        ) = proposed_block.into_parts();
+
+        let prev_block_commitment = prev_block_header.hash();
+
+        // Get the root of the block note tree.
+        // --------------------------------------------------------------------------------------------
+
+        // TODO: Do we need the full tree in proposed block or would the root be sufficient? We
+        // can reconstruct the tree from the output note batches, so the question is whether it
+        // should be included in the proven block to be readily accessible or if recomputing it is
+        // fine.
+        let note_root = block_note_tree.root();
```

Comment:
I think just the root is sufficient. In general, I think of a `ProvenBlock` as a fairly compact structure that does not contain data that can be derived from other fields in the struct.